CRUD for employees + Escalate tickets

tina262317 opened the pull request.

tina262317 committed:
CRUD for employees + Escalate tickets

tina262317 committed:
CRUD for employees + Escalate tickets

OrestPokotylenko committed:
employee list optimized (#8)

* employee list optimized

* GetAllEmployees improved

* updated DisplayEmployees

* fixes

tina262317 committed:
Added username, password and fixed roles

tina262317 committed:
Added username, password and fixed roles

tina262317 committed:
Fixed a typo

tina262317 committed:
Update after conflicts

tina262317 committed:
Merge branch 'crud-employee' of https://github.com/AnastasiaIU/the-garden-group into crud-employee

tina262317 committed:
Updated Code clean up

AnastasiaIU reviewed (changes requested):
The project is not running and the API feature is broken. Please fix it.

tina262317 committed:
Removed duplicate

tina262317 committed:
added the imports

added the imports that were deleted due to code clean up

tina262317 committed:
Updated and fixed some issues

Fixed:
-the password and username method
-ignore openTickets
-user management issues

tina262317 committed:
Fixed user mangement

AnastasiaIU commented on DAL/EmployeeDao.cs:
```diff
@@ -1,14 +1,61 @@
 {
     public class EmployeeDao : BaseDao
     {
-        public async Task<List<Employee>> GetAllEmployeesAsync()
```
Please, keep this method. It's needed for the API feature.
```
public async Task<List<Employee>> GetAllEmployeesAsync()
{
    var filter = Builders<Employee>.Filter.Empty;
    return await employeeCollection.Find(filter).ToListAsync();
}
```

AnastasiaIU commented on DAL/EmployeeDao.cs:
```diff
@@ -19,5 +66,51 @@ public async Task<List<Employee>> GetAllEmployeesAsync()
+        public async Task DeleteEmployeeByID(string employeeId)
+        {
+            // create the filter to find the employee by thier Id
+            var deleteFilter = Builders<Employee>.Filter.Eq("_id", new ObjectId(employeeId));
```
This code repeats 3 times. Please, make a separate method for getting the filter.

tina262317 replied:
Thank you that's a really good suggestion

AnastasiaIU commented on DAL/EmployeeDao.cs:
```diff
@@ -19,5 +66,51 @@ public async Task<List<Employee>> GetAllEmployeesAsync()
+                .Set("email", updatedEmployee.Email)
+                .Set("phone_number", updatedEmployee.PhoneNumber)
+                .Set("role", updatedEmployee.Role.ToString())
+                .Set("branch", updatedEmployee.Branch);
```
I believe this code is redundant.
Check this link for example:
https://stackoverflow.com/questions/30893012/update-all-properties-of-object-in-mongodb

AnastasiaIU commented on Service/EmployeeService.cs:
```diff
@@ -9,9 +9,9 @@ public class EmployeeService
     {
         private readonly EmployeeDao employeeDao = new();
 
-        public async Task<List<Employee>> GetAllEmployeesAsync()
```
Please keep the method below for the API.
```
public async Task<List<Employee>> GetAllEmployeesAsync()
{
    return await employeeDao.GetAllEmployeesAsync();
}
```

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -1,5 +1,8 @@
 using Service;
+using System.Drawing;
+using System.Media;
+using System.Numerics;
```
These imports are not used. Please remove it.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+            if (currentEmployee.Role == EmployeeRole.ServiceDeskEmployee)
+            {
+                btnEditEmployee.Visible = true;
+                btnAddEmployee.Visible = true;
```
This is unnecessary. Just hide the menu item.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+                btnEscalate.Visible = false;
+                menuItemUsers.Visible = false;
+                usersList.FullRowSelect = false;
+                ticketsListView.FullRowSelect = false;
```
Please explain the purpose of these.

tina262317 replied:
While menuItemUsers becomes invisible, the tickets list view is still visible, but the regular should not be able to select a ticket and do operations on it, including escalation. So the option to select a row becomes disabled and the escalate button becomes invisible.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+            if (usersList.SelectedItems.Count > 0)
+            {
+                ChangeButtonState(btnEditEmployee, Color.Orange, true);
+                ChangeButtonState(btnAddEmployee, Color.LightGray, false);
```
Why disable the add user button? If I choose an employee, I cannot deselect and add a user anymore.

tina262317 replied:
You can deselect an employee. Clicking inside the users view but not on a filled row will deselect an employee. I can send a video showing that. I had decided that it makes more sense for the add button to be disabled, it doesn't make sense to add an employee while selecting another employee. I can think of a few options if you think Dan might find it confusing how deselecting a row works:
- I can either put a button that says "deselect" to deselect everything
- I can put a tool-tip over disabled buttons that shows how to enable it, which would be the way to deselect for an add button (I think this could be the most useful for all buttons and makes our app better quality. It would be a simple string parameter)
- Completely ignore this and make adding a new employee possible even when there's already an employee selected
Let me know which one you think would be better for the app.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+        }
         private async void menuItemUsers_Click(object sender, EventArgs e)
+        {
+            ShowUsersView();
```
This method is only used on your part, so good job on not duplicating the code.

Please, remove the comment.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+        //{
+        //    ShowPanel(pnlTicketsOverview);
+        //    await DisplayTicketsAsync(currentEmployee);
+        //}
```
This method is not duplicated on the main, so you duplicated it during the conflict resolution.

Please remove the comment.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+
+        #endregion
+
+        // Tina's comment: are these two methods not supposed to be inside a region?
```
It is inside the User Management Logic region on the main. Please fix it how it was.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+            {
+                Employee newEmployee = CreateEmployee();
+                await employeeService.CreateEmployeeAsync(newEmployee);
+                // create a method 
```
You do not create any methods here. Please, remove this comment or change it to
`// show confirmation message`

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+            string lastName = IsInputBoxEmpty(textBoxLastName.Text, "The last name cannot be empty");
+            string email = IsInputBoxEmpty(textBoxEmailAddress.Text, "The e-mail address cannot be empty");
+            string phoneNumber = IsInputBoxEmpty(textBoxPhoneNumber.Text, "The phone number cannot be empty");
+            // first making sure the combo box is not empty then parsing it
```
I don't think small moments like that need to be explained. If you want to provide this explanation to the teacher, please make it in the document at the final sprint.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+        private Employee CreateEmployee()
+        {
+            // check for each input if they're empty
+            string firstName = IsInputBoxEmpty(textBoxFirstName.Text, "The first name cannot be empty");
```
Keep string in the resources and call them with this method:
![image](https://github.com/user-attachments/assets/cc2a6f96-62a7-4ee2-a5b5-703813d3fab0)
![image](https://github.com/user-attachments/assets/2b87f973-daa8-4621-a7b3-3d08e9b3780a)

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+            // set the selected employee to the current selected item in the list view
+            selectedEmployee = (Employee)usersList.SelectedItems[0].Tag;
+
+            // refresh the students panel so new information shows up
```
Comments in this method look odd. Please check them out.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+
+            ShowPanel(pnlAddEditUser);
+        }
+        private async void btnUpdateEmployee_Click(object sender, EventArgs e)
```
Most of your methods are crowded, please add new lines between methods for better readability.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+        }
 
+        // method to change the enablement and coloration of the buttons
+        // Do you want this method to be outside of this region if others' also need to change a button or something?
```
Remove this comment. We can discuss things in the chat. Don't need to make a chat from the code.

AnastasiaIU commented on UI/GardenDeskForm.cs:
```diff
@@ -101,27 +92,293 @@ private async void BtnLogin_Click(object sender, EventArgs e)
+            MessageBox.Show($"'{selectedTicket.Title}' has been escalated.");
+
+            // refresh the list of tickets again
+            // I suggest making a method like ShowUsersView for Tickets as well
```
Remove this comment pls.

AnastasiaIU reviewed (changes requested):
Please fix your code with accordance to the comments.

tina262317 committed:
Updated and fixed issues

Updated methods and comments and improved readability

AnastasiaIU reviewed (approved):
Good job on the improvements! The code look nice and neat 👍
